[soons1] iP

soons1 opened the pull request:
# **SnowBoy**
>"I'm just a little snowman, I have a carrot for a nose. I'm happy when it's frosty, I love it when it snows!" - The Little Snowman

_SnowBoy_ is meant to be your one-stop assistant who can help you with:
- Adding to-do tasks
- Display your list of tasks
- Mark tasks as done

As an example, here's how you might use it:
1. Add a task called read book
2. Mark it as done when completed
3. Add another task called read novel
4. View your list

To try it out for yourself, download the jar file from [here](https://github.com/soons1/ip/releases/tag/v0.1) and type in your CLI:
```
java -jar ip.jar
```

The `main` method is quite simple; it looks like this:
```java
public class Main {
    public static void main(String[] args) throws IOException {
        new Duke("data/tasks.txt").run();
    }
}
```

Features (Current and TBA)
- [X] Implement various commands
- [ ] GUI

Thank you for visiting⛄

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

soons1 committed:
Updated to fit Level 0 requirements

soons1 committed:
Updated to fit Level 1 requirements

soons1 committed:
Updated to fit Level 2 requirements

soons1 committed:
Updated to fit Level 3 requirements + Minor improvement to printLst method

soons1 committed:
Updated to fit Level 4 requirements

soons1 committed:
Updated to fit TextUITesting requirements

soons1 committed:
Updated to fit Level 5 requirements

soons1 committed:
Updated to fit Level 6 requirements

soons1 committed:
Updated to fit Level 7 requirements + Added ASCII art to greet method

soons1 committed:
Merge branch 'branch-Level-7'
Merged branch-Level-7 to master to test merge commit

soons1 committed:
Updated to fit Level 8 requirements

soons1 committed:
Merge branch 'branch-Level-8'
Merged branch-Level-8 to master to test merge commit

soons1 committed:
Update to make iP more OOP

soons1 committed:
Merge branch 'A-MoreOOP'
Merge A-MoreOOP branch to master to keep project OOP-oriented

soons1 committed:
Update to include the classes into the package duke

soons1 committed:
Merge branch 'add-gradle-support'
Add Gradle support to automate tasks

soons1 committed:
Update to fit A-JUnit requirements

soons1 committed:
Update to fit Level 9 requirements

soons1 committed:
Merge branch 'branch-Level-9'
Merge branch-Level-9 to master to finish Level 9

soons1 committed:
Update to fit A-CodingStandard requirements by editing style errors

soons1 committed:
Merge branch 'A-CodingStandard'
Merge A-CodingStandard to master to merge style fixes

soons1 committed:
Update to fit A-JavaDoc requirements by adding JavaDocs

soons1 committed:
Merge branch 'A-JavaDoc'
Merge A-JavaDoc to master to add JavaDocs

soons1 committed:
Update to fit A-CheckStyle requirements and fix checkstyle errors

soons1 committed:
Merge branch 'A-CheckStyle'
Merge A-CheckStyle to master to fix checkstyle errors and add config files

soons1 committed:
Update to fit Level-10 requirements and refactor Ui and Parser

soons1 committed:
Merge branch 'branch-Level-10'
Merge branch-Level-10 to master to update SnowBoy into a GUI

soons1 committed:
Update to fit A-Varargs requirements and fix the find functionality

soons1 committed:
Merge branch 'A-Varargs'
Merge A-Varargs to master to update Varargs use in Launcher and Parser

gavingoh99 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,67 @@
+     */
+    @Override
+    public String toString() {
+        return "[D]" + super.getStatus() + " " + super.toString() + " (by: "
+                + this.by.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
I like that you segmented your code here so that it would not be a long one-liner!

gavingoh99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,229 @@
+     * @param tl The TaskList containing the tasks to be printed.
+     */
+    public static String printLst(TaskList tl) {
+        StringBuilder toPrint = new StringBuilder();
```
I like that you used a StringBuilder rather than using Java's native string concatenation as it greatly simplifies the code as well as makes it more efficient!

gavingoh99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,79 @@
+        } catch (IOException e) {
+            throw new DukeException("Sorry, I can't find the file to load from");
+        }
+        while (s.hasNext()) {
```
Perhaps abstracting the logic within this loop into a method would be a good idea!

gavingoh99 commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,71 @@
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+
+/**
+ * Controller for MainWindow. Provides the layout for the other controls.
+ */
```
I like that you included javadocs for the new javafx classes!

gavingoh99 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,89 @@
+ * It provides functionalities to add, delete, retrieve, and search for tasks within the list.
+ */
+public class TaskList {
+    private final ArrayList<Task> lst;
```
Perhaps a more contextual naming could be used for your ArrayList here and initialize the field!
```suggestion
    private final ArrayList<Task> tasks = new ArrayList<>();
```

gavingoh99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,65 @@
+     * @param cmd The user input command to be processed.
+     * @throws DukeException If the command is invalid or incorrectly used.
+     */
+    public static String checkCmd(TaskList tl, String cmd) throws DukeException {
```
Perhaps more contextual argument names
```suggestion
    public static String checkCmd(TaskList tasks, String command) throws DukeException {
```

gavingoh99 reviewed (commented):
Overall, your efforts in documenting your code well through javadocs is evident! My only comment for the PR would be to try to scale back on the use of acronyms like cmd and tl in favour of their full names so as to reduce the chances for confusion when discussing the codebase with other programmers!

soons1 committed:
Add assertions to ensure output from fromSaveFormat method is valid

Assertions made in Storage check for the number of parameters before
calling the various fromSaveFormat method

Assertions made in the respective Task subclasses check if the isDone
parameter is valid

soons1 committed:
Improve code quality to fit the requirements of A-CodeQuality

Java Enumerations is used to parse the list of valid commands
Long and complex methods are extracted to increase readability

soons1 committed:
Merge pull request #2 from soons1/A-CodeQuality

Improve code quality to fit the requirements of A-CodeQuality

soons1 committed:
Merge branch 'A-Assertions'

Resolve merge conflicts with master branch

soons1 committed:
Add help command to display help page to user

Create help page to guide user on commands to use
Direct user to help page in greeting message

soons1 committed:
Merge branch 'C-Help'
Merge branch C-Help to master to add help command

lynnetteeee commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,64 @@
+    /**
+     * Marks the task as not done.
+     */
+    public void markAsUndone() {
```
Good that these two related methods are named similarly (i.e. markAs...)!

lynnetteeee commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,89 @@
+     * @param num The index of the task to retrieve.
+     * @return The Task at the specified index.
+     */
+    public Task get(int num) {
```
More specific naming could be used for the method e.g. `getTaskFromindex(int num)`, as part of using name to explain to enhance the readability of your code

lynnetteeee commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,89 @@
+     * @param keyword The string to search for in the task descriptions.
+     * @return An ArrayList of Tasks that contain the keyword.
+     */
+    public ArrayList<Task> find(String keyword) {
```
Similar to the previous example, more details could be included in the method name to better explain the purpose of the method. A suggestion would be `findTasksWithKeyword(String keyword)` !

lynnetteeee commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,84 @@
+     * Displays a greeting message to the user and shows the command for help.
+     *
+     */
+    public static String greet() {
```
Since your current `greet()`, `showLoadingError()` and `exit()` methods are all displaying in nature, showing the Ui, it would be more readable if they all have similar naming! The naming for `showLoadingError` is quite good as it captures enough details, so was thinking maybe the others can extend from it e.g.`showGreetings()`, `showExitMessage()`

lynnetteeee commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,101 @@
+     * @param tl The TaskList containing the tasks to be saved.
+     * @throws IOException If an I/O error occurs during writing to the file.
+     */
+    public void save(TaskList tl) throws IOException {
```
Maybe more details can be included in the method name, and following the convention of the `writeToSavedFile` method, perhaps this could be renamed to `saveTasksToFile`

lynnetteeee reviewed (commented):
Honestly LGTM. Mainly need to work on the naming of methods to include more details so that the names can do part of the explanation of what your code ultimately does! Overall good job!!

soons1 committed:
Update chatbot response to give it more personality

Include elements to suggest chatbot is in a cold environment

soons1 committed:
Merge branch 'A-Personality'

soons1 committed:
Update readme to include a simple user guide

soons1 committed:
Fix readme to correct command to run jar file

soons1 committed:
Fix hardcoded filepath for release on jar file

soons1 committed:
Rename Ui file to fit naming requirement

soons1 committed:
Delete UI file as name format does not fit requirements

soons1 committed:
Fix ParserTest.java to correctly test exception when params are missing

Add window title to GUI

Update Ui.png to show updated GUI

soons1 committed:
Update name of jar file to reflect chatbot name

soons1 committed:
Fix printing of errors on GUI

Fix README to update to new release
